src/keys.c

```diff
@@ -168,76 +155,104 @@ void public_key_hash(uint8_t *const hash_out,
                               hash_out,
                               HASH_SIZE));
+
     if (compressed_out != NULL) {
         memmove(compressed_out, &compressed, sizeof(*compressed_out));
     }
+
+end:
+    return error;
+}
+
+cx_err_t generate_public_key_hash(uint8_t *const hash_out,
+                                  size_t const hash_out_size,
+                                  cx_ecfp_public_key_t *compressed_out,
+                                  bip32_path_with_curve_t const *const path_with_curve) {
+    if ((hash_out == NULL) || (path_with_curve == NULL)) {
+        return CX_INVALID_PARAMETER;
+    }
+
+    cx_ecfp_public_key_t pubkey = {0};
+    cx_err_t error = CX_OK;
+
+    CX_CHECK(generate_public_key(&pubkey, path_with_curve));
+
+    CX_CHECK(public_key_hash(hash_out,
+                             hash_out_size,
+                             compressed_out,
+                             path_with_curve->derivation_type,
+                             &pubkey));
+
+end:
+    return error;
 }
 
-size_t sign(uint8_t *const out,
-            size_t const out_size,
-            derivation_type_t const derivation_type,
-            key_pair_t const *const pair,
-            uint8_t const *const in,
-            size_t const in_size) {
-    check_null(out);
-    check_null(pair);
-    check_null(in);
-
-    size_t tx = 0;
-    switch (derivation_type_to_signature_type(derivation_type)) {
-        case SIGNATURE_TYPE_ED25519: {
-            static size_t const SIG_SIZE = 64u;
-            if (out_size < SIG_SIZE) {
-                THROW(EXC_WRONG_LENGTH);
-            }
+cx_err_t sign(uint8_t *const out,
+              size_t *out_size,
+              bip32_path_with_curve_t const *const path_with_curve,
+              uint8_t const *const in,
+              size_t const in_size) {
+    if ((out == NULL) || (out_size == NULL) || (path_with_curve == NULL) || (in == NULL)) {
+        return CX_INVALID_PARAMETER;
+    }
 
-            CX_THROW(cx_eddsa_sign_no_throw(&pair->private_key,
-                                            CX_SHA512,
-                                            (uint8_t const *) PIC(in),
-                                            in_size,
-                                            out,
-                                            SIG_SIZE));
+    cx_err_t error = CX_OK;
 
-            tx += SIG_SIZE;
+    bip32_path_t const *const bip32_path = &path_with_curve->bip32_path;
+    derivation_type_t derivation_type = path_with_curve->derivation_type;
+    unsigned int derivation_mode = derivation_type_to_derivation_mode(derivation_type);
+    signature_type_t signature_type = derivation_type_to_signature_type(derivation_type);
+    cx_curve_t cx_curve = signature_type_to_cx_curve(signature_type);
+    uint32_t info;
 
+    switch (signature_type) {
+        case SIGNATURE_TYPE_ED25519: {
+            CX_CHECK(bip32_derive_with_seed_eddsa_sign_hash_256(derivation_mode,
```

Comment:
Same here , no seed. 
```suggestion
            CX_CHECK(bip32_derive_eddsa_sign_hash_256(derivation_mode,
```